Source/WebCore/Modules/mediastream/MediaStreamTrack.cpp

```diff
@@ -314,6 +320,75 @@ MediaStreamTrack::TrackCapabilities MediaStreamTrack::getCapabilities() const
     return result;
 }
 
+void MediaStreamTrack::queueAndProcessSerialAction(Function<Ref<GenericPromise>(SerialActionResolution)>&& action)
+{
+    ASSERT(isMainThread());
+    m_pendingActionQueue.append(WTFMove(action));
+    processNextSerialAction();
+}
+
+void MediaStreamTrack::processNextSerialAction()
+{
+    ASSERT(isMainThread());
+    if (m_currentSerialAction || m_pendingActionQueue.isEmpty())
+        return;
+
+    m_currentSerialAction = { m_pendingActionQueue.takeFirst() };
```

Comment:
It is not clear to me whether we need a m_pendingActionQueue.
Maybe what we need is using chainTo, let's check with Jean-Yves

Reply:
The queue is used for two types of task: applyConstraints which is using completionHandler and takePhoto that is returning a promise.
You can't take a photo unless applyConstraints is done.

if applyConstraints returned a NativePromise, we could indeed just chain the promises together, so that takePhoto would automatically occurs when applyConstraints is finished. 

We could probably rather than using a dedicated queue chain the action. End result would be the same

Reply:
I am not sure we need to wait for applyConstraints to call takePhoto.
First, normally, applyConstraints succeeds synchronously in GPUProcess, we just have to ensure that the order to applyConstraints is not following a different path than the order to takePhoto.
Second, applyConstraints is tied to a track, but you can call several applyConstraints on different tracks pointing to the same GPU source (either via track.clone() or calling getUserMedia twice).
If we have to create a queue, it would best be handled in GPUProcess by the capture source itself.